pytorch/training/docker/1.8.0/py3/cu111/Dockerfile.gpu

```diff
@@ -0,0 +1,259 @@
+ && ln -s ${CONDA_PREFIX}/bin/pip /usr/local/bin/pip3
+
+RUN conda install -c pytorch magma-cuda111==2.5.2 \
+ && conda install -c conda-forge \
+    opencv \
+ && conda install -y scikit-learn \
+    pandas \
+    h5py \
+    requests \
+    libgcc \
+ && conda clean -ya
+
+# Install libboost from source. This package is needed for smdataparallel functionality [for networking asynchronous IO].
+RUN wget --quiet https://dl.bintray.com/boostorg/release/1.73.0/source/boost_1_73_0.tar.gz \
+  && tar -xzf boost_1_73_0.tar.gz \
+  && cd boost_1_73_0 \
+  && ./bootstrap.sh \
+  && ./b2 threading=multi --prefix=${CONDA_PREFIX} -j 64 cxxflags=-fPIC cflags=-fPIC install || true \
+  && cd .. \
+  && rm -rf boost_1_73_0.tar.gz \
+  && rm -rf boost_1_73_0
+
+WORKDIR /opt/pytorch
+
+# Copy workaround script for incorrect hostname
+COPY changehostname.c /
+COPY start_with_right_hostname.sh /usr/local/bin/start_with_right_hostname.sh
+
+WORKDIR /root
+
+# Uninstall torch and torchvision before installing the custom versions from an S3 bucket
+RUN pip install --no-cache-dir -U \
+    smdebug==${SMDEBUG_VERSION} \
+    smclarify \
+    "sagemaker>=2,<3" \
+    sagemaker-experiments==0.* \
+    sagemaker-pytorch-training \
+    --no-cache-dir fastai==1.0.61 \
+    "pyyaml>=5.4,<5.5" \
+    "awscli<2" \
+    psutil \
+    Pillow \
+    scipy \
+    pybind11 \
+    click \
+    mpi4py==3.0.3 \
+    cmake==3.18.2.post1 \
+    torchnet \
+    "cryptography>3.2" \
+ && pip uninstall -y torch \
+ && pip install --no-cache-dir -U ${PT_TRAINING_URL} \
+ && pip uninstall -y torchvision \
+ && pip install --no-deps --no-cache-dir -U ${PT_TORCHVISION_URL}
+
+# install metis
+RUN wget -nv http://glaros.dtc.umn.edu/gkhome/fetch/sw/metis/${METIS}.tar.gz \
+  && gunzip -f ${METIS}.tar.gz \
+  && tar -xvf ${METIS}.tar \
+  && cd ${METIS} \
+  && apt-get update \
+  && apt-get install -y build-essential \
+  && apt-get install -y cmake \
+  && make config shared=1 \
+  && make install \
+  && cd .. \
+  && rm -rf ${METIS}.tar* \
+  && rm -rf ${METIS} \
+  && rm -rf /var/lib/apt/lists/* \
+  && apt-get clean
+
+# Install RAPIDSMemoryManager.
+# Requires cmake>=3.14.
+RUN  wget -nv https://github.com/rapidsai/rmm/archive/v${RMM_VERSION}.tar.gz \
+  && tar -xvf v${RMM_VERSION}.tar.gz \
+  && cd rmm-${RMM_VERSION} \
+  && INSTALL_PREFIX=/usr/local ./build.sh librmm \
+  && cd .. \
+  && rm -rf v${RMM_VERSION}.tar* \
+  && rm -rf rmm-${RMM_VERSION}
+
+# Install Horovod
+RUN pip uninstall -y horovod \
+ && ldconfig /usr/local/cuda-11.1/targets/x86_64-linux/lib/stubs \
+ && HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_CUDA_HOME=/usr/local/cuda-11.1 HOROVOD_WITH_PYTORCH=1 pip install --no-cache-dir horovod==${HOROVOD_VERSION} \
+ && ldconfig
+
+# Install Nvidia Apex
+RUN git clone https://github.com/NVIDIA/apex.git \
+ && cd apex \
+ && pip install -v --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" ./
```

Comment:
Please remove this. Does smd-mp still require Apex?

@yl-to @ydaiming @rahul003

Reply:
We still need it actually.

Reply:
As for the from source stuff thats because apex doesn't ship a release. It has to be from source

Reply:
From what I understand, Apex is a part of PyTorch source code starting from PT 1.7, right?